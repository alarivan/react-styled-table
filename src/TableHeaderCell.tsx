import React from 'react'
import styled from 'styled-components'
import { ReactStyledTableColumn, SortOrder } from '.'

interface StyledTHProps {
  sortable?: boolean
}

const StyledTH = styled.th`
  ${({ sortable }: StyledTHProps) => sortable && 'cursor: pointer;'}
`

export interface TableHeaderCellProps
  extends Pick<ReactStyledTableColumn, 'value' | 'label' | 'sortable'> {
  sortByValue?: string
  sortOrder: SortOrder
  onClick: (value: string) => void
}

const SortToggle = ({ sort }: { sort: SortOrder }) => {
  switch (sort) {
    case 'ascend':
      return <span>&#8595;</span>
    case 'descend':
      return <span>&#8593;</span>

    default:
      return (
        <React.Fragment>
          <span>&#8593;</span> <span>&#8595;</span>
        </React.Fragment>
      )
  }
}

const TableHeaderCell: React.FC<TableHeaderCellProps> = ({
  value,
  label,
  sortable,
  sortByValue,
  sortOrder,
  onClick
}) => {
  const handleClick = () => {
    if (sortable) onClick(value)
  }

  return (
    <StyledTH
      sortable={sortable}
      onClick={handleClick}
      role={sortable ? 'button' : undefined}
    >
      {label}
      {sortable && (
        <SortToggle sort={sortByValue === value ? sortOrder : undefined} />
      )}
    </StyledTH>
  )
}

export default TableHeaderCell
