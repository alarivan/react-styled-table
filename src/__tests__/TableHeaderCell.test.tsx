import 'jest-styled-components'
import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import TableHeaderCell, { TableHeaderCellProps } from '../TableHeaderCell'

const renderTh = (props: TableHeaderCellProps) => {
  return render(
    <table>
      <thead>
        <tr>
          <TableHeaderCell {...props} />
        </tr>
      </thead>
    </table>
  )
}

test('should render correctly with sortable true', () => {
  const onClick = jest.fn()
  const { container } = renderTh({
    value: 'name',
    label: 'Name',
    sortable: true,
    sortByValue: 'name',
    sortOrder: undefined,
    onClick
  })

  const tableCell = screen.getByText(/Name/)
  userEvent.click(tableCell)
  expect(onClick).toHaveBeenCalledTimes(1)

  expect(container.firstChild).toMatchSnapshot()
})

test('should render correctly with sortable false', () => {
  const onClick = jest.fn()
  const { container } = renderTh({
    value: 'name',
    label: 'Name',
    sortable: false,
    sortByValue: 'other',
    sortOrder: undefined,
    onClick
  })

  const tableCell = screen.getByText(/Name/)
  userEvent.click(tableCell)
  expect(onClick).toHaveBeenCalledTimes(0)

  expect(container.firstChild).toMatchSnapshot()
})

test('should render correctly with undefined sortOrder', () => {
  renderTh({
    value: 'name',
    label: 'Name',
    sortable: true,
    sortByValue: 'name',
    sortOrder: undefined,
    onClick: jest.fn()
  })

  expect(screen.queryByText(/↑/)).toBeTruthy()
  expect(screen.queryByText(/↓/)).toBeTruthy()
})

test('should render correctly with ascending sortOrder', () => {
  renderTh({
    value: 'name',
    label: 'Name',
    sortable: true,
    sortByValue: 'name',
    sortOrder: 'ascend',
    onClick: jest.fn()
  })

  expect(screen.queryByText(/↑/)).toBeFalsy()
  expect(screen.queryByText(/↓/)).toBeTruthy()
})

test('should render correctly with descending sortOrder', () => {
  renderTh({
    value: 'name',
    label: 'Name',
    sortable: true,
    sortByValue: 'name',
    sortOrder: 'descend',
    onClick: jest.fn()
  })

  expect(screen.queryByText(/↑/)).toBeTruthy()
  expect(screen.queryByText(/↓/)).toBeFalsy()
})
