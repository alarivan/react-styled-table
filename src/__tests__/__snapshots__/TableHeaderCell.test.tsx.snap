// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with sortable false 1`] = `
<table>
  <thead>
    <tr>
      <th
        class=""
        role="button"
      >
        Name
      </th>
    </tr>
  </thead>
</table>
`;

exports[`should render correctly with sortable true 1`] = `
.c0 {
  cursor: pointer;
}

<table>
  <thead>
    <tr>
      <th
        class="c0"
        role="button"
      >
        Name
        <span>
          ↑
        </span>
         
        <span>
          ↓
        </span>
      </th>
    </tr>
  </thead>
</table>
`;
